### ISSUE TITLE
add support for FunctionK

### ISSUE THREAD

**Issue by jcouyang** [finagle/finch#1044]
previously finch support any data type that can be convert to `TwitterFuture` by typeclass `ToTwitterFuture`

but since 0.25 it only support typeclass `Effect`, `Future` and `TwitterFuture`, which is not very friendly to customize data type that can be converted to `Effect` or `Future`

e.g. I compose all business logic to `Free[Program, Unit]`, in old finch I can simply define `ToTwitterFuture[Free[Program, ?]]` then finch can convert my type to Future

while in the new finch it can't

## maybe we can make this generic by using FunctionK
add the following to `Mapper.scala`
```scala
implicit def mapperToEffectOutputFunction[A, B, F[_], G[_]:Effect](f: A => F[Output[B]])(implicit cov:F~>G): Mapper.Aux[G, A, B] = new Mapper[G, A] {
    type Out = B
    def apply(e: Endpoint[G,A]) : Endpoint[G, B] = e.mapOutputAsync(a => cov.apply(f(a)))
  }

  implicit def mapperToEffectOutputValue[A, B, F[_], G[_]:Effect](f: => F[Output[B]])(implicit cov: F ~> G): Mapper.Aux[G, A, B] = new Mapper[G, A] {
    type Out = B
    def apply(e: Endpoint[G,A]) : Endpoint[G, B] = e.mapOutputAsync(a => cov.apply(f))
  }

  implicit def mapperFromKindOutputHFunction[F[_] : Effect, G[_], A, B, FN, FOB](f: FN)(implicit
    conv: G ~> F,
    ftp: FnToProduct.Aux[FN, A => FOB],
    ev: FOB <:< G[Output[B]]
  ): Mapper.Aux[F, A, B] = new Mapper[F, A] {
    type Out = B
    def apply(e: Endpoint[F,A]) : Endpoint[F, B] = e.mapOutputAsync(a => conv.apply(ev(ftp(f)(a))))
  }

```
then user can simply define how to convert their kind to some kind of `Effect`
```scala
case class MyType[A](value: A)

implicit val convert: MyType ~> IO  = Lambda[MyType ~> IO](a=>IO(a.value))
```

thus, use can define response in custom kind, not constrain to only `Effect` or `Future`

```scala
  def hello = get("hello" :: path[String]) { s: String =>
    MyType(Ok(Message(s)))
  }

```

**Comment by vkostyukov** [comment]
@jcouyang In #1042 we're discussing to having a type-class to convert from both Scala Futures and Twitter Futures into `F[_]: Effect`. Presumably you would be able to use that extension point to lift your own stuff into `F[_]: Effect`.

**Comment by sergeykolbasov** [comment]
@jcouyang the nice thing about having your own type class is that some basic implicits could live in companion object and there will be no need to do any additional movements to enable them. In our case we're talking about scala and twitter futures transformations.

**Comment by jcouyang** [comment]
@sergeykolbasov I think you probably missed my point here
I'm not saying typeclass is bad idea
I'm saying
> typeclass ToEffect[F] is basically just an *alias* for `? ~> E: Effect`

Since Finch already depends on cats, why border create your own typeclass instead of using the generic one cats already has
```scala
type ToEffect[F[_], E[_]: Effect] = F ~> E
```
> some basic implicits could live in companion object 
In the example implicit `futureToEffect` still live in Finch, user don't have to know about it
it's my custom type
```
implicit val convert: MyType ~> IO  = Lambda[MyType ~> IO](a=>IO(a.value))
```
that need to implement the typeclass `? ~> E` on my own

**Comment by sergeykolbasov** [comment]
Closed in #1045
